pkg/provisioners/gcppubsub/dispatcher/receiver/cache/cache.go

```diff
@@ -0,0 +1,174 @@
+//
+//    cache := cache.NewTTL()
+//    go cache.Start(stopCh)
+//    cachedValue := cache.Insert(key, value)
+//    cachedValue.Start()
+//    otherValue := cache.Get(otherKey)
+//    if otherValue != nil {
+//        ...use otherValue...
+//    }
+//    ...cachedValue times out and cachedValue.Stop() is called in a background goroutine...
+type TTL struct {
+	itemsLock     sync.Mutex
+	items         map[key]*item
+	ttl           time.Duration
+	reapingPeriod time.Duration
+	currentTime   func() time.Time
+}
+
+// item is stored in the cache.TTL's items map. It represents the value stored in the cache, as well
+// as the time it lives until.
+type item struct {
+	value Stoppable
+	ttl   time.Time
+}
+
+// Stoppable is the interface that wraps a basic Stop method.
+type Stoppable interface {
+	// Stop is the method called when an item is evicted from the cache.
+	Stop()
+}
+
+// NewTTL creates a new cache.TTL. Note that for the culling to occur, Start(stopCh) must be called.
+//
+//     cache := cache.NewTTL()
+//     go cache.Start(stopCh)
+func NewTTL() *TTL {
+	return &TTL{
+		itemsLock:     sync.Mutex{},
+		items:         make(map[key]*item),
+		ttl:           5 * time.Minute,
+		reapingPeriod: 5 * time.Minute,
+		currentTime:   time.Now,
+	}
+}
+
+// Insert inserts an item into the cache. If something is already stored with that key, then the
+// existing stored value is returned. This is especially important if the value needs to be started,
+// in which case the returned value must be started, not the input.
+//
+//     cachedValue := cache.Insert(key, value)
+//     // Note that cachedValue may either be value or something that was already in the cache.
+//     cachedValue.Start()
+func (c *TTL) Insert(keyString string, value Stoppable) Stoppable {
+	newItem := &item{
+		value: value,
+		ttl:   c.currentTime().Add(c.ttl),
+	}
+	k := key(keyString)
+	c.itemsLock.Lock()
+	defer c.itemsLock.Unlock()
+	current, present := c.items[k]
+	if present {
+		current.extend(c.currentTime, c.ttl)
+		return current.value
+	} else {
+		c.items[k] = newItem
+		return value
+	}
+}
+
+// Get gets an item in the cache and extends its time to live. If the key is not present in the
+// cache, nil is returned.
+func (c *TTL) Get(keyString string) Stoppable {
+	k := key(keyString)
+	c.itemsLock.Lock()
+	defer c.itemsLock.Unlock()
+	i := c.items[k]
+	if i != nil {
+		i.extend(c.currentTime, c.ttl)
+		return i.value
+	}
+	return nil
+}
+
+// extend extends the time this item's time to live to now plus ttl.
+func (i *item) extend(currentTime func() time.Time, ttl time.Duration) {
+	i.ttl = currentTime().Add(ttl)
+}
+
+// Start continuously culls the old items from the cache. It should be run in a goroutine.
+func (c *TTL) Start(stopCh <-chan struct{}) error {
+	for {
+		select {
+		case <-stopCh:
+			return nil
+		case <-time.After(c.reapingPeriod):
+			c.cull()
+		}
+	}
+}
+
+// cull culls all items from the cache that are past their time to live.
```

Comment:
learned a new word today :)

nit: seems a bit redundant, maybe just:
// cull all items from the cache
vs. cull culls ?

Reply:
Done.